lib/perso.ml

```diff
@@ -1216,6 +1239,295 @@ let warning_use_has_parents_before_parent (fname, bp, ep) var r =
+  | _, _ ->
+      let t', d_t' =
+        let max_l1, max_l2 = max_l1_l2_aux env in
+        Printf.sprintf "******** Compute %d × %d table ********\n" max_l1 max_l2
+        |> !GWPARAM.syslog `LOG_WARNING;
+        (* TODO test for Sys.max_array_length *)
+        let cousins_cnt = Array.make_matrix (max_l2 + 1) (max_l2 + 1) [] in
+        let cousins_dates = Array.make_matrix (max_l2 + 1) (max_l2 + 1) [] in
+        cousins_cnt.(0).(0) <-
+          [ (get_iper p, [ Gwdb.dummy_ifam ], Gwdb.dummy_iper, [ 0 ]) ];
+        cousins_dates.(0).(0) <- get_min_max_dates conf base cousins_cnt.(0).(0);
+        let rec loop0 j =
+          (* initiate lists of direct descendants *)
+          cousins_cnt.(0).(j) <- descendants base cousins_cnt 0 j;
+          cousins_dates.(0).(j) <-
+            get_min_max_dates conf base cousins_cnt.(0).(j);
+          if j < max_l2 then loop0 (j + 1) else ()
+        in
+        loop0 1;
+        let rec loop1 i =
+          (* get ascendants *)
+          cousins_cnt.(i).(0) <- ascendants base [] cousins_cnt.(i - 1).(0) i;
+          cousins_dates.(i).(0) <-
+            get_min_max_dates conf base cousins_cnt.(i).(0);
+          let rec loop2 i j =
+            (* get descendants of c1, except persons of previous level (c2) *)
+            cousins_cnt.(i).(j) <- descendants base cousins_cnt i j;
+            cousins_dates.(i).(j) <-
+              get_min_max_dates conf base cousins_cnt.(i).(j);
+            if j < max_l2 then loop2 i (j + 1)
+            else if i < max_l1 then loop1 (i + 1)
+            else ()
+          in
+          loop2 i 1
+        in
+        loop1 1;
+        (cousins_cnt, cousins_dates)
+      in
+      cousins_t := Some t';
+      cousins_dates_t := Some d_t';
+      (t', d_t')
+
+let max_l1_l2 conf base env p =
+  let cousins_cnt, _cousins_dates =
+    match (!cousins_t, !cousins_dates_t) with
+    | Some t, Some d_t -> (t, d_t)
+    | _, _ -> init_cousins_cnt conf base env p
+  in
+  let max_l1, max_l2 = max_l1_l2_aux env in
+  let max_i =
+    let rec loop0 i =
+      if cousins_cnt.(i).(0) <> [] && i < max_l1 then loop0 (i + 1) else i
+    in
+    loop0 0
+  in
+  let rec loop i j =
+    if cousins_cnt.(i).(j) <> [] then
+      if j < max_l2 then loop i (j + 1) else (max_i, j - i)
+    else if i < max_l1 && j < max_l2 then loop (i + 1) (j + 1)
+    else (max_i, j - i)
+  in
+  loop 0 0
+
+let cousins_l1_l2_aux conf base env l1 l2 p =
+  let il1 = int_of_string l1 in
+  let il2 = int_of_string l2 in
+  let max_l1, max_l2 = max_l1_l2_aux env in
+  if il1 <= max_l1 && il2 - il1 <= max_l2 then
+    let cousins_cnt, _cousins_dates = init_cousins_cnt conf base env p in
+    (* gros calcul *)
+    Some cousins_cnt.(il1).(il2)
+  else None
+
+(* create a new list (ip, (ifaml, iancl, cnt), lev) from (ip, ifaml, ipar, lev) *)
+let cousins_fold list =
+  let same_ifaml ifl1 ifl2 =
+    List.for_all2 (fun if1 if2 -> if1 = if2) ifl1 ifl2
+  in
+  let list = List.sort compare list in
```

Comment:
please can you use `l` for the name of variable that are of type list instead of `list`?
since `list` is a type, most text editor color it differently and it is confusing/ugly.
Also `l` is good enough to make it obvious that it is a list